classes/source/source_url.php

```diff
@@ -0,0 +1,209 @@
+        global $CFG;
+
+        parent::__construct($settings);
+
+        $this->now = date('YmdHis', time());
+
+        $this->filedir = $CFG->dataroot . DIRECTORY_SEPARATOR . $this->get_short_name();
+        check_dir_exists($this->filedir);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function extract() {
+        if ($this->is_available()) {
+            $this->data = new data([$this->get_file()]);
+        } else {
+            throw new \Exception($this->name . ' source is not available!');
+        }
+
+        return $this->data;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function is_available() {
+
+        if ($this->check_remote_url($this->get_url_path(), $this->get_local_file_path($this->get_url_path()))) {
+            return true;
+        }
+
+        $this->log('load_data', 'URL is not valid ' . $this->get_url_path(), logger::TYPE_ERROR);
+
+        return false;
+    }
+
+    /**
+     * Copy file from source.
+     *
+     * @param string $remotefile Remote file path.
+     * @param string $localfile Local file path.
+     *
+     * @return bool
+     * @throws \coding_exception
+     */
+    protected function copy_file($remotefile, $localfile) {
+        $result = copy($remotefile, $localfile);
+
+        if ($result) {
+            $this->log('copy_from_folder', 'Completed copy ' . $remotefile . ' to ' . $localfile);
+        } else {
+            $this->log('copy_from_folder', 'Failed to copy ' . $remotefile . ' to ' . $localfile, logger::TYPE_ERROR);
+        }
+
+        return $result;
+    }
+
+    /**
+     * Return url address.
+     *
+     * @return string
+     */
+    protected function get_url_path() {
+        return $this->settings['address'];
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function create_config_form_elements(\MoodleQuickForm $mform) {
+        $elements = parent::create_config_form_elements($mform);
+
+        $fields = array(
+            'address' => new config_field('address', 'URL', 'text', $this->settings['address'], PARAM_URL),
+        );
+
+        return array_merge($elements, $this->get_config_form_elements($mform, $fields));
+    }
+
+    /**
+     * Get a path of local file to copy.
+     *
+     * @param string $remotefile Remote file path.
+     *
+     * @return string
+     */
+    protected function get_local_file_path($remotefile) {
+        $localfolder = $this->filedir . DIRECTORY_SEPARATOR . $this->now;
+        check_dir_exists($localfolder);
+
+        return $localfolder . DIRECTORY_SEPARATOR . basename($remotefile);
+    }
+
+    /**
+     * Check if remote file downloadable.
+     *
+     * @param string $url Remote file url.
+     *
+     * @return int
+     */
+    protected function check_remote_url($url, $filename) {
+        global $CFG;
+
+        $c = new curl();
+
+        $result = $c->download_one($url, null, array('filepath' => $filename, 'timeout' => $CFG->repositorygetfiletimeout));
```

Comment:
repositorygetfiletimeout is related to the repository api, I would not re-use this here. If you want it I'd just make it a custom config item in this plugin with a sane default